### PR TITLE
Ramdump

### PR DESCRIPTION
**Pull request by thapav**
Update UART ramdump to memory region selective ramdump & its support for artik05x.
This patch updates the ramdump tool which dumps the whole memory together.
The updated tool will dump memory region wise based on user's input as below:
        Please enter desired ramdump option as below:
                1 for ALL
                2 for Region 0
                25 for Region 0 & 3 ...
Each region will be dumped into a seperate bin file of respective region size.

### CODE REVIEW

**Comment by seinfra** [comment]
Target : [f3c8a9a48e93aeda340b6bf0a53d7cc95050e2a2] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [f3c8a9a48e93aeda340b6bf0a53d7cc95050e2a2] - Code Rule Check Result:
os/board/artik05x/src/artik05x_crashdump.c:88: ERROR: [SPC_M_OPR] space required after that ',' (ctx:VxO)
os/board/artik05x/src/artik05x_crashdump.c:88: ERROR: [SPC_M_OPR] space required before that '*' (ctx:OxV)

File: tools/ramdump/ramdump_tool.c [hunk 1]
```diff
@@ -80,69 +88,221 @@ static int b_read(int fd, uint8_t *buf, int size)
...
+		}
+	}
+	printf("\n=========================================================================\n");
+
+	printf("Please enter desired ramdump option as below:\n \t1 for ALL\n");
+	if (number_of_regions > 1) {
+	printf(" \t2 for Region 0\n \t25 for Region 0 & 3 ...\n");
+	}
+	printf("Please enter your input: ");
+	scanf("%d", &index);
```
Comment (yashwanth686007) [inline tools/ramdump/ramdump_tool.c]: What happens if user inputs a invalid input or a negative integer.

File: tools/ramdump/ramdump_tool.c [hunk 2]
```diff
@@ -80,69 +88,221 @@ static int b_read(int fd, uint8_t *buf, int size)
...
+}
+
+static int ramdump_recv(int dev_fd)
+{
+	char buf;
+	int i;
+	int ret;
+	int index = 1;
+	int count = 0;
+	int dump_all = 0;
```
Comment (yashwanth686007) [inline tools/ramdump/ramdump_tool.c]: Same as above. I think this variable is not necessary.

File: tools/ramdump/ramdump_tool.c [hunk 3]
```diff
@@ -80,69 +88,221 @@ static int b_read(int fd, uint8_t *buf, int size)
...
-
-	printf("%s: ramdump_address = %08x, ramdump_size = %d\n", __func__, ramdump_address, ramdump_size);
-
-	snprintf(bin_file, BINFILE_NAME_SIZE, "ramdump_0x%08x_0x%08x.bin", ramdump_address, (ramdump_address + ramdump_size));
-
-	bin_fp = fopen(bin_file, "w");
-	if (bin_fp == NULL) {
-		printf("%s create failed\n", bin_file);
+	/* Array to store memory region size */
+	rd_regionx_size = malloc((number_of_regions + 2) * sizeof(int));
```
Comment (yashwanth686007) [inline tools/ramdump/ramdump_tool.c]: Its better to typecast the malloc calls.

File: tools/ramdump/ramdump_tool.c [hunk 4]
```diff
@@ -80,69 +88,221 @@ static int b_read(int fd, uint8_t *buf, int size)
...
+	scanf("%d", &index);
+
+	if (index == 1) {
+		rd_regionx_map[index] = 0;
+		for (i = 2; i <= number_of_regions + 1; i++) {
+			rd_regionx_map[i] = 1;
+			regions_to_dump++;
+		}
+	}
+	else if (index != 1) {
```
Comment (jeongchanKim) [inline tools/ramdump/ramdump_tool.c]: I think only else is enough.

File: tools/ramdump/ramdump_tool.c [hunk 5]
```diff
@@ -80,69 +88,221 @@ static int b_read(int fd, uint8_t *buf, int size)
...
+		/* Map index values asked to dump */
+		while(index > 0) {
```
Comment (jeongchanKim) [inline tools/ramdump/ramdump_tool.c]: Please add space after `while`

File: tools/ramdump/ramdump_tool.c [hunk 6]
```diff
@@ -80,69 +88,221 @@ static int b_read(int fd, uint8_t *buf, int size)
...
+			rd_regionx_map[index%10] = 1;
+			index = index/10;
```
Comment (jeongchanKim) [inline tools/ramdump/ramdump_tool.c]: Please add space before and after '/'

File: tools/ramdump/ramdump_tool.c [hunk 7]
```diff
@@ -80,69 +88,221 @@ static int b_read(int fd, uint8_t *buf, int size)
...
-		printf("%s create failed\n", bin_file);
+	/* Array to store memory region size */
+	rd_regionx_size = malloc((number_of_regions + 2) * sizeof(int));
+	if (!rd_regionx_size) {
+		return -1;
+	}
+	/* Array to store heap index corresponding to memory regions */
+	rd_regionx_idx = malloc((number_of_regions + 2) * sizeof(int));
+	if (!rd_regionx_idx) {
+		return -1;
```
Comment (jeongchanKim) [inline tools/ramdump/ramdump_tool.c]: Is it okay not to free the `rd_regionx_size`?

File: tools/ramdump/ramdump_tool.c (hunk 1 above)
Comment (shadaksharikp) [reply tools/ramdump/ramdump_tool.c]: user input needs to be validated before the usage

File: tools/ramdump/ramdump_tool.c [hunk 8]
```diff
@@ -80,69 +88,221 @@ static int b_read(int fd, uint8_t *buf, int size)
...
+
+	return 0;
+}
+
+static int ramdump_recv(int dev_fd)
+{
+	char buf;
+	int i;
+	int ret;
+	int index = 1;
```
Comment (shadaksharikp) [inline tools/ramdump/ramdump_tool.c]: if the num_of_regions are not more than 255, you can use unsigned char for i & index variable at multiple places used above. Hence multiple bytes can be saved to achieve space efficient code.

**Comment by seinfra** [comment]
Target : [52875aeb9143477d585fbca3eef36d319c81f9ba] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [52875aeb9143477d585fbca3eef36d319c81f9ba] - Code Rule Check (C++) OK.

File: tools/ramdump/ramdump_tool.c [hunk 9]
```diff
@@ -80,69 +91,202 @@ static int b_read(int fd, uint8_t *buf, int size)
...
 		return -1;
 	}
 
-	printf("%s: ramdump_address = %08x, ramdump_size = %d\n", __func__, ramdump_address, ramdump_size);
+	/* Receive memory address, size & heap index for memory regions from TARGET */
+	for (i = 2; i <= number_of_regions + 1; i++) {
+		ret = b_read(dev_fd, (uint8_t *)&mem_address, 4);
+		if (ret != 4) {
+			printf("Receiving address failed, ret = %d\n", ret);
+			return -1;
```
Comment (jeongchanKim) [inline tools/ramdump/ramdump_tool.c]: Is it okay not to free `mem_info`?

File: tools/ramdump/ramdump_tool.c [hunk 10]
```diff
@@ -34,6 +35,16 @@
...
 typedef unsigned char uint8_t;
 
+/* Ramdump initialization data */
+uint32_t  number_of_regions;
+typedef struct {
+	int rd_regionx_idx;
+	uint32_t rd_regionx_start;
+	uint32_t rd_regionx_size;
+	int rd_regionx_mark;
+}rd_regionx;
```
Comment (jeongchanKim) [inline tools/ramdump/ramdump_tool.c]: Please add space after '}'.

File: tools/ramdump/ramdump_tool.c [hunk 11]
```diff
@@ -80,69 +91,202 @@ static int b_read(int fd, uint8_t *buf, int size)
...
+	if (ret != 0) {
+		printf("Receiving number of regions to be dumped failed, ret = %d\n", ret);
+		return -1;
+	}
+	printf("\nNo. of Regions to be dumped received: %s\n", __func__);
+	printf("\nReceiving ramdump......\n\n");
+
+	/* Dump data region wise */
+	for (index = 2; index <= number_of_regions + 1; index++) {
+		if ( mem_info[index].rd_regionx_mark == 1) {
```
Comment (jeongchanKim) [inline tools/ramdump/ramdump_tool.c]: Please remove space after '('.

File: tools/ramdump/ramdump_tool.c [hunk 12]
```diff
@@ -80,69 +91,202 @@ static int b_read(int fd, uint8_t *buf, int size)
...
+					printf("Writing ramdump %dTH byte failed, ret = %d\n", count, ret);
+					fclose(bin_fp);
+					return -1;
+				}
+
+				count++;
+				ramdump_size--;
+
+				if ((count % (16 * 1024)) == 0) {
+				//	printf("%s: received %d bytes\n", __func__, count);
```
Comment (jeongchanKim) [inline tools/ramdump/ramdump_tool.c]: If not needed, please remove this line.

**Comment by seinfra** [comment]
Target : [a435f96bfb04c7934cdf3a59c41541865ee143ce] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [a435f96bfb04c7934cdf3a59c41541865ee143ce] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [37ed1d458f385c725d75dfe0a928b7937965c100] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [37ed1d458f385c725d75dfe0a928b7937965c100] - Code Rule Check (C++) OK.

File: tools/ramdump/ramdump_tool.c [hunk 13]
```diff
@@ -80,69 +91,201 @@ static int b_read(int fd, uint8_t *buf, int size)
...
-			printf("Writing ramdump %dTH byte failed, ret = %d\n", count, ret);
-			fclose(bin_fp);
-			return -1;
+	/* Take user's input for desired ramdump region */
+	printf("Please enter your input: ");
+	scanf("%d", &index);
+
+scan_input:
+	/* Check if user's input is valid */
+	while (index < 0 || index == EOF || (number_of_regions == 1 && index > 1)) {
```
Comment (yashwanth686007) [inline tools/ramdump/ramdump_tool.c]: If the number of regions is 1, can't "2" be passed as a parameter?

Comment (thapav) [reply tools/ramdump/ramdump_tool.c]: no, because that's not valid, so there will be a correct input prompt until user enters correct options

**Comment by thapav** [comment]
@manoharahk @sunghan-chang  Please review

File: tools/ramdump/ramdump_tool.c [hunk 14]
```diff
@@ -80,69 +91,201 @@ static int b_read(int fd, uint8_t *buf, int size)
...
+				if (ret != 1) {
+					printf("Writing ramdump %dTH byte failed, ret = %d\n", count, ret);
+					fclose(bin_fp);
+					return -1;
+				}
+
+				count++;
+				ramdump_size--;
+
+				if ((count % (16 * 1024)) == 0) {
```
Comment (jeongchanKim) [inline tools/ramdump/ramdump_tool.c]: How about changing 16 and 1024 to meaningful defined value?

File: tools/ramdump/ramdump_tool.c [hunk 15]
```diff
@@ -80,69 +91,201 @@ static int b_read(int fd, uint8_t *buf, int size)
...
+scan_input:
+	/* Check if user's input is valid */
+	while (index < 0 || index == EOF || (number_of_regions == 1 && index > 1)) {
+		printf("Please enter correct input: ");
+		scanf("%d", &index);
+	}
+
+	/* Mark regions to be dumped */
+	while (index > 0) {
+		if((index % 10) > (number_of_regions + 1)) {
```
Comment (jeongchanKim) [inline tools/ramdump/ramdump_tool.c]: Please add space after 'if'.

File: tools/ramdump/ramdump_tool.c (hunk 14 above)
Comment (thapav) [reply tools/ramdump/ramdump_tool.c]: done

File: tools/ramdump/ramdump_tool.c (hunk 15 above)
Comment (thapav) [reply tools/ramdump/ramdump_tool.c]: done

**Comment by seinfra** [comment]
Target : [a7e8b2dd883bb44f0339940fede1e4228f7cc5f4] -

**Comment by seinfra** [comment]
Target : [a7e8b2dd883bb44f0339940fede1e4228f7cc5f4] -

**Comment by seinfra** [comment]
Target : [c72e637b6c8daee7769bd4b13d8000ceaf2c796c] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [c72e637b6c8daee7769bd4b13d8000ceaf2c796c] - Code Rule Check OK.

**Pull request merged**